### PR TITLE
Improve weight compression time and memory for BF16 OV models

### PR DESCRIPTION
**Pull request by nikita-savelyevv**
### Changes

Avoid retrieving weight data if not necessary.

For llama-3-8b with BF16 weights there is about 1.5x time and 1.26x peak memory improvement. Please see the figure below.
| Before         | After   |
|-------------------|-------|
| system_memory_usage_from-zero |  system_memory_usage_from-zero |

### Reason for changes

For BF16 dtype when constant data is fetched, it is loaded into memory resulting in additional time and memory overhead.

### CODE REVIEW

File: nncf/quantization/algorithms/weight_compression/openvino_backend.py [hunk 1]
```diff
@@ -106,6 +106,28 @@ def get_weight(self, node_with_weight: NNCFNode, weight_port_id: int, model: ov.
         weight_tensor = get_const_value(weight_node)
         return Tensor(weight_tensor)
 
+    def get_weight_dtype(
+        self, node_with_weight: NNCFNode, weight_port_id: int, model: ov.Model, graph: NNCFGraph
+    ) -> TensorDataType:
+        weight_name = node_with_weight.layer_attributes.constant_attributes[weight_port_id]["name"]
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/openvino_backend.py:112]: I think suboptimal constant type definition may be a common problem for NNCF. What do you think about adding dtype to constant_attributes and reusing it in the algorithm?

File: nncf/quantization/algorithms/weight_compression/torch_backend.py [hunk 2]
```diff
@@ -170,6 +170,16 @@ def get_weight(
...
         return Tensor(weight)
 
+    def get_weight_dtype(
+        self, node_with_weight: NNCFNode, weight_port_id: int, model: torch.nn.Module, graph: NNCFGraph
+    ) -> TensorDataType:
+        return self.get_weight(node_with_weight, weight_port_id, model, graph).dtype
+
+    @staticmethod
+    def get_weight_shape(node_with_weight: NNCFNode, weight_port_id: int, graph: NNCFGraph) -> Tuple:
+        weight_node = graph.get_previous_nodes(node_with_weight)[weight_port_id]
```
Comment (alexsu52) [inline nncf/quantization/algorithms/weight_compression/torch_backend.py:180]:
```suggestion
        weight_node = get_const_node(node_with_weight, weight_port_id, graph)
```

**Review by alexsu52 — COMMENTED** [review]
Good stuff!

File: nncf/quantization/algorithms/weight_compression/openvino_backend.py (hunk 1 above)
Comment (nikita-savelyevv) [reply nncf/quantization/algorithms/weight_compression/openvino_backend.py]: Added dtype to OV node constant_attributes

File: nncf/quantization/algorithms/weight_compression/torch_backend.py (hunk 2 above)
Comment (nikita-savelyevv) [reply nncf/quantization/algorithms/weight_compression/torch_backend.py]: Thanks, changed

**Review by alexsu52 — APPROVED** [review]
LGTM

**Comment by nikita-savelyevv** [comment]
> Did you run `post_training_weight_compression_performance`? It would be useful to see how we compare on this scope with previous behavior.

Thanks for the suggestion. Will run. Do you know, are there models with bf16 weights?

**Comment by ljaljushkin** [comment]
> > Did you run `post_training_weight_compression_performance`? It would be useful to see how we compare on this scope with previous behavior.
> 
> Thanks for the suggestion. Will run. Do you know, are there models with bf16 weights?

Don't think we can run on SPR via CI. maybe run locally on it?

**Comment by nikita-savelyevv** [comment]
I did not notice any changes in time or memory of the `post_training_weight_compression_performance` build (id 14) against the results for develop (id 15). As I understand this is because there `compress_weights()` is called on a model with weights in FP32 precision. At least I see that an original tiny-llama model contains only fp32 constants. It is also saved into `fp32_models` directory.

I also ran `post_training_weight_compression` job (id 158). There are no regressions.

**Review by ljaljushkin — APPROVED** [review]

**Pull request merged**